core/libs/shared/src/lazy.rs

```diff
@@ -181,10 +184,10 @@ impl<T: Stagable> LazyTree<T> {
         Ok(name)
     }
 
-    pub fn link(&self, id: &Uuid) -> SharedResult<Option<Uuid>> {
+    pub fn link(&mut self, id: &Uuid) -> SharedResult<Option<Uuid>> {
         for link_id in self.owned_ids() {
             if let FileType::Link { target } = self.find(&link_id)?.file_type() {
-                if id == &target {
+                if id == &target && !self.calculate_deleted(&link_id)? {
```

Comment:
this call to calculate deleted necessitates a mut reference which necessitates slight re-arrangement of `in_pending_share` above